py/objmodule.c

```diff
@@ -239,45 +239,47 @@ STATIC const mp_rom_map_elem_t mp_builtin_module_table[] = {
 
 MP_DEFINE_CONST_MAP(mp_builtin_module_map, mp_builtin_module_table);
 
-// returns MP_OBJ_NULL if not found
-mp_obj_t mp_module_get(qstr module_name) {
+// Returns an already-loaded module (of any type), or MP_OBJ_NULL otherwise.
+mp_obj_t mp_module_get_loaded(qstr module_name) {
     mp_map_t *mp_loaded_modules_map = &MP_STATE_VM(mp_loaded_modules_dict).map;
     // lookup module
     mp_map_elem_t *el = mp_map_lookup(mp_loaded_modules_map, MP_OBJ_NEW_QSTR(module_name), MP_MAP_LOOKUP);
+    return el ? el->value : MP_OBJ_NULL;
+}
 
+// Returns an already-loaded module, a newly-initialised built-in, or MP_OBJ_NULL otherwise.
+mp_obj_t mp_module_get(qstr module_name) {
+    mp_obj_t loaded = mp_module_get_loaded(module_name);
+    if (loaded != MP_OBJ_NULL) {
+        // Already loaded, so avoid double-calling init
+        return loaded;
+    }
+    // Look for builtin module names
+    mp_map_elem_t *el = mp_map_lookup((mp_map_t*)&mp_builtin_module_map, MP_OBJ_NEW_QSTR(module_name), MP_MAP_LOOKUP);
     if (el == NULL) {
-        // module not found, look for builtin module names
-        el = mp_map_lookup((mp_map_t*)&mp_builtin_module_map, MP_OBJ_NEW_QSTR(module_name), MP_MAP_LOOKUP);
-        if (el == NULL) {
-            return MP_OBJ_NULL;
-        }
-        mp_module_call_init(module_name, el->value);
+        return MP_OBJ_NULL;
     }
+    mp_module_call_init(module_name, el->value);
 
     // module found, return it
     return el->value;
 }
 
+#if MICROPY_MODULE_WEAK_LINKS
+// Same as mp_module_get, but prefixes
+mp_obj_t mp_module_get_builtin_umodule(qstr module_name) {
+    char buf[MICROPY_ALLOC_PATH_MAX];
```

Comment:
To avoid this stack allocation (which can be large) it might be better to pass in a buffer from the caller.  Eg the `VSTR_FIXED(path, MICROPY_ALLOC_PATH_MAX)` (not sure if that's allowed to be clobbered tho at the point of calling this function).

Reply:
Or even pass in the vstr itself and use vstr functions here to construct the u-module name.

Reply:
I reused the path vstr. I'd prefer to do this while the path is empty (i.e. two lines higher) but then you end up creating a new "umodule" QSTR for every import.